### ISSUE TITLE
Internal error in Model differences view when unfolding some items while comparing models

### ISSUE THREAD

**Issue by eposse** [ZeligsoftDev/CX4CBDDS#132]
# **Issue and tracking information**

### Developer's Actual time spent on fix (hours)  

# **Issue reporter to provide a detailed description of the issue in the space below**

When using Papyrus Compare to compare models, if you unfold some elements with references to DDS4CCM model elements (in particular CORBAException elements) a pop-up appears indicating a java.lang.NullPointerException and one or more error messages. These appear in the Error Log as well.

#### Steps to reproduce

(This assumes that you have the CX4CBDDS git repository added in the git perspective and the 'papyrus' branch checked-out).

1. Import the BasicPubSub_new project from the CX4CBDDS git repo 'examples' folder into the workspace
1. Open the BasicPubSub_new model (BasicPubSub/ModelFiles/BasicPubSubPapyrus)
1. Open the assembly diagram (In the Model Explorer navigate to /BasicPubSub/assemblies/BasicPubSub_asm/BasicPubSub_asm and double click on the "NewCompositeStructureDiagram" item to open it)
1. Add a new subscriber part (In the Model Explorer navigate to BasicPubSub/components/SNA_Examples/BasicSubscriber_comp, drag the BasicSubscriber_comp interface and drop it into the diagram).
1. Save the model.
1. Open the Git perspective (Window > Perspective > Open Perspective > Other... > Git)
1. Go to the "Git Staging" view.
1. Stage the /BasicPubSub/ModelFiles/BasicPubSubPapyrus.notation and the BasicPubSub/ModelFiles/BasicPubSubPapyrus.uml files. You may need to click the "Refresh" button in the top-right of the "Git Staging" view.
1. Go back to the Papyrus perspective (Window > Perspective > Open Perspective > Other... > Papyrus)
1. Open the Error Log view (Window > Show View > Error Log)
1. Delete the current log entries (right-click on the "Delete Log" button, a red [X] on the top-right of the Error Log view).
1. Right click on /BasicPubSub/ModelFiles/BasicPubSubPapyrus and select "Compare With > Index)
1. If prompted to switch to the Synchronization perspective, click on [Yes] (and optionally check the "Remember my decision" checkbox)
1. In the Synchronization view on the left, unfold BasicPubSub/ModelFiles, right-click on BasicPubSubPapyrus.notation and select "Open in Compare Editor".
1. In the "Model Differences" view on top, navigate to /BasicPubSub/ports/SNA_Examples/BasicPubSub/TestData_conn/Reader/query, unfolding items as necessary.
1. Unfold the last item "query".

#### Expected result

The Model Differences view displays the contents or references of the 'query' item.

#### Actual result

A pop-up appears with NPE errors, several exceptions appear in the Error Log and the Model Differences view shows "Internal Error" in place of the contents of the 'query' item. 

See the attached log.

[exceptions.log](https://github.com/ZeligsoftDev/CX4CBDDS/files/4623906/exceptions.log)

**Comment by eposse** [comment]
The main exception is this NPE:

```
java.lang.NullPointerException
	at org.eclipse.emf.edit.ui.provider.ComposedImageDescriptor.getSize(ExtendedImageRegistry.java:407)
	at org.eclipse.jface.resource.CompositeImageDescriptor.getImageData(CompositeImageDescriptor.java:369)
	at org.eclipse.swt.graphics.Image.<init>(Image.java:820)
	at org.eclipse.jface.resource.ImageDescriptor.createImage(ImageDescriptor.java:293)
	at org.eclipse.jface.resource.ImageDescriptor.createImage(ImageDescriptor.java:242)
	at org.eclipse.jface.resource.ImageDescriptor.createImage(ImageDescriptor.java:220)
	at org.eclipse.emf.edit.ui.provider.ExtendedImageRegistry.getImage(ExtendedImageRegistry.java:185)
	at org.eclipse.emf.edit.ui.provider.AdapterFactoryLabelProvider.getImageFromObject(AdapterFactoryLabelProvider.java:409)
	at org.eclipse.emf.edit.ui.provider.AdapterFactoryLabelProvider.getImage(AdapterFactoryLabelProvider.java:380)
	at org.eclipse.emf.compare.ide.ui.internal.structuremergeviewer.EMFCompareStructureMergeViewerLabelProvider.getImage(EMFCompareStructureMergeViewerLabelProvider.java:117)
...
```

It seems like the cause of this is that the label provider is unable to get an icon for the `CORBAException` element in the DDS4CCM.uml model.

**Comment by eposse** [comment]
The function at the heart of the problem is this:

```java
	@Override
	public Object getImage(Object object) {
		ReferenceChange refChange = (ReferenceChange)object;

		Object refChangeValueImage = itemDelegator.getImage(refChange.getValue());

		Object diffImage = overlayProvider.getComposedImage(refChange, refChangeValueImage);
		Object ret = overlayImage(object, diffImage);

		return ret;
	}
```

The call to `itemDelegator.getImage(refChange.getValue());` returns `null` because:

1) Down the line it calls `org.eclipse.papyrus.infra.ui.emf.providers.StandardEMFLabelProvider.getImage(EObject)` which tries to obtain an "ItemLabelProvider" (in function `org.eclipse.papyrus.infra.ui.emf.providers.StandardEMFLabelProvider.getItemLabelProvider(EObject)` for the target `EObject` which is an element from the DDS4CCM.uml model. More precisely it is `org.eclipse.emf.ecore.impl.DynamicEObjectImpl@1d28ce39 (eProxyURI: pathmap://DDS4CCM_DCPS_CONNECTOR_LIBRARIES/DDS_DCPS.uml#_XRXKIGROEd-Wc9hRqhM12g eClass: org.eclipse.emf.ecore.impl.EClassImpl@4b6d8d39 (name: CORBAException) (instanceClassName: null) (abstract: false, interface: false))`. 

2) To obtain the label provider, it tries to obtain an "itemProviderAdapterFactory" for the URI `http://www.zeligsoft.com/domain/dds4ccm/2010/DDS4CCM/1`, but no such factory has been registered (through the `org.eclipse.emf.edit.itemProviderAdapterFactories` extension point), so the factory is `null`.

3) Since the factory is `null`, so is the label provider, and this which ultimately leads to `refChangeValueImage` in the function above to be `null`.

4) Then, the call to `overlayImage` returns an object representing the overlay image with an array of images, but the first one actually `null`. 

5) So eventually it goes back to `org.eclipse.emf.edit.ui.provider.ExtendedImageRegistry.getImage(Object)` line 185, where it tries to compose the images by calling `composedImageDescriptor.createImage();` which eventually crashes when it gets to `org.eclipse.jface.resource.CompositeImageDescriptor.getImageData(int)` and tries to get the size of the `null` image.

**Comment by eposse** [comment]
There is a partial solution:

In com.zeligsoft.domain.dds4ccm/plugin.xml register the following extension:

```xml
   <extension
         point="org.eclipse.emf.edit.itemProviderAdapterFactories">
      <factory
            class="org.eclipse.uml2.uml.edit.providers.UMLItemProviderAdapterFactory"
            supportedTypes="org.eclipse.emf.edit.provider.IItemLabelProvider"
            uri="http://www.zeligsoft.com/domain/dds4ccm/2010/DDS4CCM/1">
      </factory>
   </extension>   
```

This basically associates the DDS4CCM model to the ItemProviderAdapterFactory from UML.

This allows `org.eclipse.papyrus.infra.ui.emf.providers.StandardEMFLabelProvider.getImage(EObject)` to resolve the factory and therefore the label provider, giving a non-null image and therefore eliminating the NPE.

Nevertheless, it doesn't work perfectly: first, the icon for the CORBAException objects looks wrong, as if it was not a valid image. Second, the text label is still `Internal Error`.

**Comment by eposse** [comment]
New information: The labels (icon + text) shown in the structural tree of the Model Differences view come from Papyrus's `LabelProviderService` (`org.eclipse.papyrus.infra.services.labelprovider.service.impl.LabelProviderServiceImpl`). 

The service maintains a list of "contexts" and can associate different label providers to different contexts, but if there are no contexts, there is a default provider called `SharedExtensibleLabelProvider` (`org.eclipse.papyrus.infra.services.labelprovider.service.impl.LabelProviderServiceImpl.SharedExtensibleLabelProvider`) which keeps a list of nested label providers, each one associated to a particular model.

These nested providers are registered with the `SharedExtensibleLabelProvider` by regisitering them with the `LabelProviderService` via an extension point: `org.eclipse.papyrus.infra.services.labelprovider.labelProvider`. This extension point requires two things: a priority and a provider class implementing the `IFilteredLabelProvider` interface.

When creating the labels, the Model Differences view delegates the label creation to the default `SharedExtensibleLabelProvider` which inherits from `ExtensibleLabelProvider` its nested providers and the `getImage` method:

```java
	@Override
	public Image getImage(Object element) {
		return getProvider(element).getImage(element);
	}
```

As can be seen here, it first tries to get a provider for the element and then delegate the `getImage` message to it. 

The `getProvider` method will ask each of the `IFilteredLabelProvider` if it "accepts" the given element. In most cases this is a UML `Element` and the associated provider is the `UMLFilteredLabelProvider` (`org.eclipse.papyrus.uml.tools.providers.UMLFilteredLabelProvider`), which knows how to obtain icons from UML elements or their stereotypes.

The problem in this issue is that the element in question (a `CORBAException`) is actually a dynamic `EObject`, i.e. a `org.eclipse.emf.ecore.impl.DynamicEObjectImpl` and more specifically a proxy object with a pathmap reference to the actual model element (`org.eclipse.emf.ecore.impl.DynamicEObjectImpl@1d28ce39 (eProxyURI: pathmap://DDS4CCM_DCPS_CONNECTOR_LIBRARIES/DDS_DCPS.uml#_XRXKIGROEd-Wc9hRqhM12g eClass: org.eclipse.emf.ecore.impl.EClassImpl@4b6d8d39 (name: CORBAException) (instanceClassName: null) (abstract: false, interface: false))`), and the `UMLFilteredLabelProvider` does not accept this type of element, so a different label provider is selected for it, thus resulting in an unexpected icon and label.

**Comment by eposse** [comment]
More info. 

`UMLFilteredLabelProvider` doesn't accept a DynamicEObject that is a proxy, e.g. `org.eclipse.emf.ecore.impl.DynamicEObjectImpl@22ca1225 (eProxyURI: pathmap://DDS4CCM_DCPS_CONNECTOR_LIBRARIES/DDS_DCPS.uml#_XRXKIGROEd-Wc9hRqhM12g eClass: org.eclipse.emf.ecore.impl.EClassImpl@12f67c01 (name: CORBAException) (instanceClassName: null) (abstract: false, interface: false))`.

This leads to the `EMFFilteredLabelProvider` (`org.eclipse.papyrus.infra.ui.emf.providers.EMFLabelProvider`) accepting the object, as it accepts every EMF model element (any `EObject` that is not a `Resource`).

This one forwards the `getImage` message to `org.eclipse.papyrus.infra.ui.emf.providers.ForwardingEMFLabelProvider` which is a subclass of `org.eclipse.papyrus.infra.ui.emf.providers.StandardEMFLabelProvider`.

This in turn, obtains an `IItemLabelProvider` from an EditFactory which is an `AdapterFactory`. It gets the factory from an extension point called `org.eclipse.emf.edit.itemProviderAdapterFactories`. Namely, the extension is selected if its 'uri' attribute it matches the URI of the EObject's eResource. In this case the uri is 
`http://www.zeligsoft.com/domain/dds4ccm/2010/DDS4CCM/1`.

The previous solution defined such extension with 'class' attribute 
`org.eclipse.uml2.uml.edit.providers.UMLItemProviderAdapterFactory`, which is then adapted to `IItemLabelProvider` which is actually a `org.eclipse.uml2.uml.edit.providers.StereotypeApplicationItemProvider`.

This `IItemLabelProvider` produces the following URI for the icon: `bundleentry://315.fwk959869407/icons/full/obj16/Item.gif#CORBAException`, from which a `org.eclipse.emf.edit.ui.provider.URLImageDescriptor.URLImageDescriptor` is created with `CORBAException` as key. The result is that the "Item.gif" icon from EMF is used instead of the one for `CORBAException`.

**Comment by eposse** [comment]
Fixed by pull request #135.
